src/hooks/useArrowKeyFocusManager.ts

```diff
@@ -45,51 +52,134 @@ export default function useArrowKeyFocusManager({
         [isActive, shouldExcludeTextAreaNodes],
     );
 
-    useEffect(() => onFocusedIndexChange(focusedIndex), [focusedIndex, onFocusedIndexChange]);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    useEffect(() => onFocusedIndexChange(focusedIndex), [focusedIndex]);
 
     const arrowUpCallback = useCallback(() => {
         if (maxIndex < 0) {
             return;
         }
+        const nextIndex = disableCyclicTraversal ? -1 : maxIndex;
 
         setFocusedIndex((actualIndex) => {
-            const currentFocusedIndex = actualIndex > 0 ? actualIndex - 1 : maxIndex;
+            let currentFocusedIndex = -1;
+            if (allowHorizontalArrowKeys) {
+                currentFocusedIndex = actualIndex > 0 ? actualIndex - itemsPerRow : nextIndex;
+            } else {
+                currentFocusedIndex = actualIndex > 0 ? actualIndex - 1 : nextIndex;
+            }
             let newFocusedIndex = currentFocusedIndex;
 
             while (disabledIndexes.includes(newFocusedIndex)) {
-                newFocusedIndex = newFocusedIndex > 0 ? newFocusedIndex - 1 : maxIndex;
+                newFocusedIndex -= allowHorizontalArrowKeys ? itemsPerRow : 1;
+                if (newFocusedIndex < 0) {
+                    break;
+                }
                 if (newFocusedIndex === currentFocusedIndex) {
                     // all indexes are disabled
                     return actualIndex; // no-op
                 }
             }
             return newFocusedIndex;
         });
-    }, [disabledIndexes, maxIndex]);
+    }, [allowHorizontalArrowKeys, disableCyclicTraversal, disabledIndexes, itemsPerRow, maxIndex]);
     useKeyboardShortcut(CONST.KEYBOARD_SHORTCUTS.ARROW_UP, arrowUpCallback, arrowConfig);
 
     const arrowDownCallback = useCallback(() => {
         if (maxIndex < 0) {
             return;
         }
 
+        const nextIndex = disableCyclicTraversal ? maxIndex : 0;
```

Comment:
can this be declared outside of this callback?

Reply:
decided to leave it as it is after internal discussion